Add compiler plugin continuation to project settings

kasper-f opened the pull request:
I have today made a fork and implemented a solution for adding continuation compiler plugin to idea config. See bug https://github.com/mpeltonen/sbt-idea/issues/102

Only problem is that I am not sure the path is calculated the correct way or if I somehow could get access to the "addCompilerPlugin" lib defined in the build. The above solution works for me on linux.

Do you see any problems including this in please let me know and I will try solving it

kasper-f committed:
Add compiler plugin continuation to project settings

mpeltonen commented:
Thanks for the patch. I think it would be better to be more general here however, and not restrict this to continuations plugin.

It looks like you can see all the plugins in scalacOptions if you look it in Compile scope:

``` code
> show compile:scalac-options
[info] List(-P:continuations:enable, -Xplugin:/Users/mpe/.ivy2/cache/org.scala-lang.plugins/continuations/jars/continuations-2.10.0.jar, -Xplugin:/Users/mpe/.ivy2/cache/org.scala-tools.sxr/sxr_2.9.0/jars/sxr_2.9.0-0.2.7.jar)
```

project.scalacOptions could be changed to use Compile scope like this in SbtIdeaPlugin:

``` code
-    val scalacOptions: Seq[String] = settings.optionalTask(Keys.scalacOptions).getOrElse(Seq())
+    val scalacOptions: Seq[String] = settings.optionalTask(Keys.scalacOptions in Compile).getOrElse(Seq())
```

This way you could loop all scalacOptions that start with -Xplugin, parse the path from there and add all plugins with correct path to xml.

This also means you have to filter out -Xplugin -options when setting scalacOptions to Idea config (IdeaModuleDescriptor line 157).

Hope this helps.

kasper-f committed:
Changed implementation acording to Mikko Peltonen guides in the comment on the pull request.
(https://github.com/mpeltonen/sbt-idea/pull/203#issuecomment-13190607)

Extracted the scalacOptions mapping to a map and added translation of "-explanintypes", "-optimise" and "-nowarn".

The defined mappings will no longer be inclueded in the compiler options tag as idea don't do this.

All paths with -Xplugin: will be include in the plugin path list

kasper-f commented:
Thanks for the comments, I am pretty new SBT so still learning;-) 

I have tried to implement it as you suggested but have also made some other changes to the scala facet generation:

I have separated the mapping from scalacOption to xml tag into a single map and then uses the map to map the entire scalacOption seq with. 

I filter out options defined in idea from the idea compilerOption option (see note in the diff)

Hope it fits better now

kasper-f commented on src/main/scala/org/sbtidea/IdeaModuleDescriptor.scala:
```diff
@@ -149,12 +158,25 @@ class IdeaModuleDescriptor(val imlDir: File, projectRoot: File, val project: Sub
+          }
+        }
+        {
+        val options=nonXplugin.filter(x=> !optionMap.contains(x)).mkString(" ")
```
uses the seq where xPlugin is filtered out and filteres out all options defined by other idea options defined in the optionMap

jozic commented on src/main/scala/org/sbtidea/IdeaModuleDescriptor.scala:
```diff
@@ -149,12 +158,25 @@ class IdeaModuleDescriptor(val imlDir: File, projectRoot: File, val project: Sub
+            <option name="pluginPaths">
+              <array>
+                {
+                project.scalacOptions.filter(x=> x.startsWith("-Xplugin:")).map(x=>x.substring(9)).
```
can "xplugin" be used here instead?

kasper-f committed:
Use the xplugin val to generate pluginPaths thanks to Eugene Platonov for pointing that out.

kasper-f committed:
Merge branch 'sbt-0.12' of https://github.com/mpeltonen/sbt-idea into sbt-0.12

mpeltonen commented:
I cherry-picked your changes, added a test and made some minor code cleanups. Test if 1.4.0-SNAPSHOT works for you or if you find some issues!

Thanks for the patch!
